presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/FailureAwareHiveMetaStoreClient.java

```diff
@@ -0,0 +1,359 @@
+            throws TException
+    {
+        return runWithHandle(() -> delegate.getPartitionsByNames(databaseName, tableName, partitionNames));
+    }
+
+    @Override
+    public List<Role> listRoles(String principalName, PrincipalType principalType)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.listRoles(principalName, principalType));
+    }
+
+    @Override
+    public PrincipalPrivilegeSet getPrivilegeSet(HiveObjectRef hiveObject, String userName, List<String> groupNames)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.getPrivilegeSet(hiveObject, userName, groupNames));
+    }
+
+    @Override
+    public List<HiveObjectPrivilege> listPrivileges(String principalName, PrincipalType principalType, HiveObjectRef hiveObjectRef)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.listPrivileges(principalName, principalType, hiveObjectRef));
+    }
+
+    @Override
+    public List<String> getRoleNames()
+            throws TException
+    {
+        return runWithHandle(delegate::getRoleNames);
+    }
+
+    @Override
+    public boolean grantPrivileges(PrivilegeBag privilegeBag)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.grantPrivileges(privilegeBag));
+    }
+
+    @Override
+    public boolean revokePrivileges(PrivilegeBag privilegeBag)
+            throws TException
+    {
+        return runWithHandle(() -> delegate.revokePrivileges(privilegeBag));
+    }
+
+    @Override
+    public void setUGI(String userName)
+            throws TException
+    {
+        runWithHandle(() -> delegate.setUGI(userName));
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return this.delegate.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        FailureAwareHiveMetaStoreClient other = (FailureAwareHiveMetaStoreClient) obj;
+        return Objects.equals(this.delegate, other.delegate);
+    }
+
+    public interface ResponseHandle
+    {
+        void success();
+
+        void failed(TException e);
+    }
+
+    public static final ResponseHandle NO_OPS_RESPONSE_HANDLE = new ResponseHandle()
+    {
+        @Override
+        public void success()
+        {
+        }
+
+        @Override
+        public void failed(TException e)
+        {
+        }
+    };
+
+    private interface ThrowingSupplier<T>
+    {
+        T get()
+                throws TException;
+    }
+
+    private interface ThrowingRunnable
+    {
+        void run()
+                throws TException;
+    }
+
+    private <T> T runWithHandle(ThrowingSupplier<T> supplier)
+            throws TException
+    {
+        try {
+            T t = supplier.get();
+            responseHandle.success();
```

Comment:
nit: move this after `catch`, so that `.success()` failure doesn't trigger `failed()`

```
T t;
try {
            t = supplier.get();
}
catch (TException e) {
            responseHandle.failed(e);
            throw e;
}
responseHandle.success();
return t;
```